File: src/lightweight-root/backends/github_actions.py
"""
GitHub Actions adapter for workflow triggers
"""
import requests
from typing import Dict, Optional, Any
from .base import CloudAdapter

class GitHubActionsAdapter(CloudAdapter):
    """Trigger GitHub Actions workflows"""
    
    def deploy(self, image_tag: str, env_vars: Optional[Dict[str, str]] = None, **kwargs) -> Dict[str, Any]:
        """Trigger GitHub workflow dispatch"""
        
        repo = self.config['repository']  # format: owner/repo
        workflow = self.config['workflow']  # e.g., worker.yml
        token = self.config['token']
        
        url = f"https://api.github.com/repos/{repo}/actions/workflows/{workflow}/dispatches"
        headers = {
            'Authorization': f'Bearer {token}',
            'Accept': 'application/vnd.github.v3+json'
        }
        
        payload = {
            'ref': self.config.get('ref', 'main'),
            'inputs': {
                'image_tag': image_tag,
                'env_vars': str(env_vars or {}),
                **(kwargs.get('inputs', {}))
            }
        }
        
        response = requests.post(url, headers=headers, json=payload)
        response.raise_for_status()
        
        return {
            'id': f"{repo}:{workflow}:{image_tag}",
            'status': 'triggered',
            'platform': 'github_actions'
        }
    
    def get_status(self, deployment_id: str) -> Dict[str, Any]:
        """Get workflow run status (simplified)"""
        return {
            'id': deployment_id,
            'status': 'running',
            'platform': 'github_actions'
        }

File: src/execution-gateway/providers/githubActions.js
const { requestJson } = require('../lib/http');
const { GatewayError } = require('../lib/errors');

class GitHubActionsClient {
  constructor(config, logger, globalTimeoutMs = 15000) {
    this.config = config || {};
    this.logger = logger;
    this.timeoutMs = globalTimeoutMs;
  }

  headers() {
    if (!this.config.token) throw new GatewayError(500, 'GitHub token is not configured');
    return {
      'Content-Type': 'application/json',
      'Accept': 'application/vnd.github+json',
      'Authorization': `Bearer ${this.config.token}`,
      'X-GitHub-Api-Version': '2022-11-28',
    };
  }

  url(pathname, params = {}) {
    const base = (this.config.api_base || 'https://api.github.com').replace(/\/$/, '');
    const u = new URL(base + pathname);
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== null) u.searchParams.set(k, String(v));
    }
    return u.toString();
  }

  repoPath(suffix) {
    if (!this.config.owner || !this.config.repo) {
      throw new GatewayError(500, 'GitHub owner/repo not configured');
    }
    return `/repos/${this.config.owner}/${this.config.repo}${suffix}`;
  }

  async submit(task = {}) {
    const workflow = task.workflow_id || task.workflow;
    const ref = task.ref || 'main';

    if (!workflow) throw new GatewayError(400, 'workflow_id is required for GitHub Actions');

    await requestJson(
      this.url(this.repoPath(`/actions/workflows/${workflow}/dispatches`)),
      {
        method: 'POST',
        headers: this.headers(),
        body: JSON.stringify({ ref, inputs: task.inputs || {} }),
      },
      this.timeoutMs,
    );

    let externalId = task.run_id || null;
    if (!externalId) {
      const { payload } = await requestJson(
        this.url(this.repoPath(`/actions/workflows/${workflow}/runs`), {
          per_page: 1,
          event: 'workflow_dispatch',
        }),
        { method: 'GET', headers: this.headers() },
        this.timeoutMs,
      );
      externalId = payload?.workflow_runs?.[0]?.id || null;
    }

    return {
      external_id: externalId,
      status: 'submitted',
    };
  }

  async getStatus(externalId) {
    if (!externalId) return { status: 'submitted', provider_status: 'unknown' };

    const { payload } = await requestJson(
      this.url(this.repoPath(`/actions/runs/${externalId}`)),
      { method: 'GET', headers: this.headers() },
      this.timeoutMs,
    );

    let status = 'running';
    if (payload?.status === 'completed') {
      status = payload.conclusion === 'success' ? 'completed' : 'failed';
    }

    return {
      status,
      provider_status: payload?.status,
      conclusion: payload?.conclusion,
      details: payload,
    };
  }

  async getLogs(externalId) {
    if (!externalId) throw new GatewayError(400, 'GitHub run id is required for logs');

    const resp = await fetch(this.url(this.repoPath(`/actions/runs/${externalId}/logs`)), {
      method: 'GET',
      headers: this.headers(),
      redirect: 'manual',
    });

    // GitHub 常见返回 302，并在 location 中给出 zip 下载链接。
    if (resp.status === 302 || resp.status === 301) {
      return {
        logs: 'GitHub logs are a zip archive. Use log_download_url to download.',
        log_download_url: resp.headers.get('location'),
      };
    }

    if (!resp.ok) {
      const text = await resp.text();
      throw new GatewayError(resp.status, `GitHub logs request failed: ${resp.status}`, text.slice(0, 1000));
    }

    const text = await resp.text();
    return {
      logs: text,
      log_download_url: null,
    };
  }

  async cancel(externalId) {
    if (!externalId) throw new GatewayError(400, 'GitHub run id is required for cancel');

    const { payload } = await requestJson(
      this.url(this.repoPath(`/actions/runs/${externalId}/cancel`)),
      { method: 'POST', headers: this.headers() },
      this.timeoutMs,
    );

    return { status: 'cancelled', details: payload };
  }
}

module.exports = { GitHubActionsClient };
